web/src/lib.rs

```diff
@@ -1286,6 +1167,116 @@ impl RuffleInstance {
     }
 }
 
+impl Drop for RuffleInstance {
+    fn drop(&mut self) {
+        self.canvas.remove();
+
+        // Stop all audio playing from the instance.
+        let _ = self.with_core_mut(|core| {
+            core.audio_mut().stop_all_sounds();
+            core.flush_shared_objects();
+        });
+
+        // Clean up all event listeners.
+        if let Some(mouse_move_callback) = self.mouse_move_callback.take() {
```

Comment:
(Nit/Idea: It might not be viable or a good idea, but, I feel like one could abstract away most of these statements.

Like, if you stored the different data in an array like this (not valid Rust code):

```rust
let event_listener_description = [
    (self.mouse_move_callback, self.canvas, "pointermove"),
    (self.mouse_enter_callback, self.canvas, "pointerenter"),
    (self.mouse_leave_callback, self.canvas, "pointerleave"),
    ...
]
```

And then implemented a method that iterated over it, like a `foreach` or similar, that might look like this (not valid Rust code):

```rust
event_listener_description.foreach(|(callback, target, event)| {
       if let Some(callb) = callback.take() {
            target
                .remove_event_listener_with_callback(
                    event,
                    callb.as_ref().unchecked_ref(),
                )
                .warn_on_error();
        }
})
```

Though, even if there may be merit to this refactoring, it might be dangerous to do now, and it might be better suited to a later PR. For instance, it would not be possible with just 3 parameters to do with `window_mouse_down_callback`, since it passes in an extra `true` argument, and during a refactoring, that might be lost. Care would at the very least be required.

Please close this comment once you have read it).

Reply:
Mike suggested this a while ago (in https://github.com/ruffle-rs/ruffle/pull/2505#pullrequestreview-568626038):
> I think better is to make an explicit `EventDispatcher`/`EventListener` struct in Rust that organizes a lot of this code up, and can have a `Drop` impl that auto-removes the listener.

Reply:
(Did you mean to reply to https://github.com/ruffle-rs/ruffle/pull/10958#discussion_r1186915849 ?)

Reply:
I had the same thought when touching that code, but I prefer to do that in a separate PR.